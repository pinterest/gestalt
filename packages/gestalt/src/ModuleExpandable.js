// @flow strict
import React, { useState, useEffect, type Node } from 'react';
import Box from './Box.js';
import Divider from './Divider.js';
import ModuleExpandableItem from './ModuleExpandableItem.js';
import { type ExpandableBaseProps } from './moduleTypes.js';

export default function ModuleExpandable({
  id,
  accessibilityExpandLabel,
  accessibilityCollapseLabel,
  expandedIndex,
  onExpandedChange,
  items,
}: ExpandableBaseProps): Node {
  const [expandedId, setExpandedId] = useState<?number>(
    Number.isFinite(expandedIndex) ? expandedIndex : null,
  );

  useEffect(() => {
    setExpandedId(Number.isFinite(expandedIndex) ? expandedIndex : null);
  }, [expandedIndex, setExpandedId]);

  return (
    <Box rounding={2} borderStyle="shadow">
      {items.map(({ icon, iconAccessibilityLabel, title, type, summary, children }, index) => (
        <React.Fragment key={index}>
          <Box>
            <ModuleExpandableItem
              id={`${id}-${index}`}
              title={title}
              icon={icon}
              iconAccessibilityLabel={iconAccessibilityLabel}
              summary={summary}
              isCollapsed={expandedId !== index}
              type={type}
              accessibilityExpandLabel={accessibilityExpandLabel}
              accessibilityCollapseLabel={accessibilityCollapseLabel}
              onModuleClicked={(isExpanded) => {
                if (onExpandedChange) {
                  onExpandedChange(isExpanded ? null : index);
                }
                setExpandedId(isExpanded ? null : index);
              }}
            >
              {children}
            </ModuleExpandableItem>
          </Box>
          {index !== items.length - 1 && <Divider />}
        </React.Fragment>
      ))}
    </Box>
  );
}
