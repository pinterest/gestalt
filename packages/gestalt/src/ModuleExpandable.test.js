// @flow strict
import React from 'react';
import renderer from 'react-test-renderer';
import ModuleExpandable from './ModuleExpandable.js';

describe('Module Expandable', () => {
  test('renders correctly with one item', () => {
    const tree = renderer
      .create(
        <ModuleExpandable
          id="uniqueTestID"
          accessibilityExpandLabel="click to expand"
          accessibilityCollapseLabel="click to collapse"
          items={[
            {
              title: 'Title',
              summary: ['summary1', 'summary2', 'summary3'],
              children: 'Children',
            },
          ]}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders correctly with multiple items', () => {
    const tree = renderer
      .create(
        <ModuleExpandable
          id="uniqueTestID"
          accessibilityExpandLabel="click to expand"
          accessibilityCollapseLabel="click to collapse"
          items={[
            {
              title: 'Title1',
              summary: ['summary1'],
              children: 'Children1',
            },
            {
              title: 'Title2',
              summary: ['summary2'],
              children: 'Children2',
            },
            {
              title: 'Title3',
              summary: ['summary3'],
              children: 'Children3',
              type: 'error',
            },
          ]}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

test('renders correctly with multiple items with expandedId', () => {
  const tree = renderer
    .create(
      <ModuleExpandable
        id="uniqueTestID"
        accessibilityExpandLabel="click to expand"
        accessibilityCollapseLabel="click to collapse"
        expandedIndex={0}
        onExpandedChange={() => {}}
        items={[
          {
            title: 'Title1',
            summary: ['summary1'],
            children: 'Children1',
          },
          {
            title: 'Title2',
            summary: ['summary2'],
            children: 'Children2',
          },
          {
            title: 'Title3',
            summary: ['summary3'],
            children: 'Children3',
            type: 'error',
          },
        ]}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
