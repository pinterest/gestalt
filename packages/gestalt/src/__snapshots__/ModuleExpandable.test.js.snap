// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with multiple items 1`] = `
<div
  className="box rounding2 shadow"
>
  <div
    className="box"
  >
    <div
      aria-controls="uniqueTestID-0"
      aria-disabled={false}
      aria-expanded={false}
      aria-label="click to expand"
      className="hideOutline tapTransition rounding0 accessibilityOutline fullWidth pointer"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="box paddingX6 paddingY6 xsDisplayFlex"
      >
        <div
          className="box flexGrow itemsBaseline marginEnd6 xsDisplayFlex"
        >
          <div
            className="box xsCol6 xsDisplayFlex"
          >
            <div
              className="Text fontSize3 darkGray alignLeft breakWord fontWeightBold truncate"
              title="Title1"
            >
              Title1
            </div>
          </div>
          <div
            className="box marginStart6 xsCol6"
          >
            <div
              className="box marginBottom0"
            >
              <div
                className="Text fontSize2 darkGray alignLeft breakWord fontWeightNormal truncate"
                title="summary1"
              >
                summary1
              </div>
            </div>
          </div>
        </div>
        <div
          className="box"
          id="uniqueTestID-0"
        >
          <div
            className="box paddingX1 paddingY1"
          >
            <svg
              aria-hidden={null}
              aria-label="click to expand"
              className="icon darkGray iconBlock"
              height="12"
              role="img"
              viewBox="0 0 24 24"
              width="12"
            >
              <path
                d="test-file-stub"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr
    className="divider"
  />
  <div
    className="box"
  >
    <div
      aria-controls="uniqueTestID-1"
      aria-disabled={false}
      aria-expanded={false}
      aria-label="click to expand"
      className="hideOutline tapTransition rounding0 accessibilityOutline fullWidth pointer"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="box paddingX6 paddingY6 xsDisplayFlex"
      >
        <div
          className="box flexGrow itemsBaseline marginEnd6 xsDisplayFlex"
        >
          <div
            className="box xsCol6 xsDisplayFlex"
          >
            <div
              className="Text fontSize3 darkGray alignLeft breakWord fontWeightBold truncate"
              title="Title2"
            >
              Title2
            </div>
          </div>
          <div
            className="box marginStart6 xsCol6"
          >
            <div
              className="box marginBottom0"
            >
              <div
                className="Text fontSize2 darkGray alignLeft breakWord fontWeightNormal truncate"
                title="summary2"
              >
                summary2
              </div>
            </div>
          </div>
        </div>
        <div
          className="box"
          id="uniqueTestID-1"
        >
          <div
            className="box paddingX1 paddingY1"
          >
            <svg
              aria-hidden={null}
              aria-label="click to expand"
              className="icon darkGray iconBlock"
              height="12"
              role="img"
              viewBox="0 0 24 24"
              width="12"
            >
              <path
                d="test-file-stub"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr
    className="divider"
  />
  <div
    className="box"
  >
    <div
      aria-controls="uniqueTestID-2"
      aria-disabled={false}
      aria-expanded={false}
      aria-label="click to expand"
      className="hideOutline tapTransition rounding0 accessibilityOutline fullWidth pointer"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="box paddingX6 paddingY6 xsDisplayFlex"
      >
        <div
          className="box flexGrow itemsBaseline marginEnd6 xsDisplayFlex"
        >
          <div
            className="box xsCol6 xsDisplayFlex"
          >
            <div
              className="box marginEnd2"
            >
              <svg
                aria-hidden={true}
                aria-label=""
                className="icon red iconBlock"
                height={16}
                role="img"
                viewBox="0 0 24 24"
                width={16}
              >
                <path
                  d="test-file-stub"
                />
              </svg>
            </div>
            <div
              className="Text fontSize3 red alignLeft breakWord fontWeightBold truncate"
              title="Title3"
            >
              Title3
            </div>
          </div>
          <div
            className="box marginStart6 xsCol6"
          >
            <div
              className="box marginBottom0"
            >
              <div
                className="Text fontSize2 darkGray alignLeft breakWord fontWeightNormal truncate"
                title="summary3"
              >
                summary3
              </div>
            </div>
          </div>
        </div>
        <div
          className="box"
          id="uniqueTestID-2"
        >
          <div
            className="box paddingX1 paddingY1"
          >
            <svg
              aria-hidden={null}
              aria-label="click to expand"
              className="icon darkGray iconBlock"
              height="12"
              role="img"
              viewBox="0 0 24 24"
              width="12"
            >
              <path
                d="test-file-stub"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with one item 1`] = `
<div
  className="box rounding2 shadow"
>
  <div
    className="box"
  >
    <div
      aria-controls="uniqueTestID-0"
      aria-disabled={false}
      aria-expanded={false}
      aria-label="click to expand"
      className="hideOutline tapTransition rounding0 accessibilityOutline fullWidth pointer"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="box paddingX6 paddingY6 xsDisplayFlex"
      >
        <div
          className="box flexGrow itemsBaseline marginEnd6 xsDisplayFlex"
        >
          <div
            className="box xsCol6 xsDisplayFlex"
          >
            <div
              className="Text fontSize3 darkGray alignLeft breakWord fontWeightBold truncate"
              title="Title"
            >
              Title
            </div>
          </div>
          <div
            className="box marginStart6 xsCol6"
          >
            <div
              className="box marginBottom2"
            >
              <div
                className="Text fontSize2 darkGray alignLeft breakWord fontWeightNormal truncate"
                title="summary1"
              >
                summary1
              </div>
            </div>
            <div
              className="box marginBottom2"
            >
              <div
                className="Text fontSize2 darkGray alignLeft breakWord fontWeightNormal truncate"
                title="summary2"
              >
                summary2
              </div>
            </div>
            <div
              className="box marginBottom0"
            >
              <div
                className="Text fontSize2 darkGray alignLeft breakWord fontWeightNormal truncate"
                title="summary3"
              >
                summary3
              </div>
            </div>
          </div>
        </div>
        <div
          className="box"
          id="uniqueTestID-0"
        >
          <div
            className="box paddingX1 paddingY1"
          >
            <svg
              aria-hidden={null}
              aria-label="click to expand"
              className="icon darkGray iconBlock"
              height="12"
              role="img"
              viewBox="0 0 24 24"
              width="12"
            >
              <path
                d="test-file-stub"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
